src/FluentValidation/Validators/OnFailureValidator.cs

```diff
@@ -0,0 +1,79 @@
+﻿using FluentValidation.Internal;
+using FluentValidation.Results;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace FluentValidation.Validators
+{
+	public class OnFailureValidator<T, TProperty> : NoopPropertyValidator
+	{
+		private readonly IPropertyValidator _innerValidator;
+		private readonly Action<T> _onFailureSuperSimple;

```

Comment:
Let's not have all these different delegate fields. Let's have just 1 that's an `Action<PropertyValidatorContext>`.  The others can all be supported using extension methods that provide delegates that wrap this one (various other methods in FV work in the same way)

Reply:
I need those 3 fields because I can pass 3 different actions:
`Action<T>`
`Action<T, PropertyValidatorContext>`
`Action<T, PropertyValidatorContext, string>`

How can I store those 3 types as one?

Reply:
Instead of defining a new delegate, you could use C# 7's support for local functions to make it a bit nicer.

Reply:
@JeremySkinner  something like this?

```
public class OnFailureValidator<T, TProperty> : NoopPropertyValidator {
	private readonly IPropertyValidator _innerValidator;
	private readonly Action<T, PropertyValidatorContext, string> _onFailure;

	public OnFailureValidator(IPropertyValidator innerValidator, Action<T, PropertyValidatorContext, string> onFailure) {
		_innerValidator = innerValidator;
		_onFailure = onFailure;
	}

	public OnFailureValidator(IPropertyValidator innerValidator, Action<T> onFailure)
		: this(innerValidator, (instance, ctx) => onFailure(instance)) {
	}

	public OnFailureValidator(IPropertyValidator innerValidator, Action<T, PropertyValidatorContext> onFailure)
		: this(innerValidator, (instance, context, errorMessage) => onFailure(instance, context)) {
	}

	public override IEnumerable<ValidationFailure> Validate(PropertyValidatorContext context) {
		var results = _innerValidator.Validate(context);
		if (!results.Any()) return results;
		var errorMessage = results.First().ErrorMessage;
		_onFailure((T)context.Instance, context, errorMessage);
		return results;
	}

	public override Task<IEnumerable<ValidationFailure>> ValidateAsync(PropertyValidatorContext context, CancellationToken cancellation) {
		var results = _innerValidator.ValidateAsync(context, cancellation);
		if (!results.Result.Any()) return results;
		var errorMessage = results.Result.First().ErrorMessage;
		_onFailure((T)context.Instance, context, errorMessage);
		return results;
	}
}
```

Reply:
Yes but let's do it in the extension method rather than using constructor chaining for consistency with the rest of the framework. 

Also your async method is blocking. You shouldn't be using `results.Result` at any point. Use await instead.

Reply:
@JeremySkinner first part fixed :)
If I'll use await then ValidateAsync will have async modifier.

With all those changes my code is getting simpler and simpler :)